op-batcher: remove `ThrottleInterval` and split block loading and batch publishing into separate goroutines

geoknee opened the pull request:
This moves the hybrid "event + interval" model for DA throttling to a purely event driven model. There are two events which can trigger throttling: (1) a change in the amount of pending data building up in the batcher and (2) a change in the active sequencer which the batcher is pointing at. 

A happy side effect is that the batcher's loading, publishing and throttling workloads are broken up and now work fully asynchronously. Previously the batcher would load, and then publish for up to some maximum amount of time (blocking block loading and throttling), then return to loading again in order to trigger throttling.

* Closes #12838 
* Replaces some use of a `for / select` loop with `ctx.Done()` branch with a simpler `range` over a channel. This makes shutdown behaviour easier to reason about and makes deadlocks less likely.
* Reduces the use of global variables, preferring local ones (wait groups, contexts, etc)
* Readme has a diagram showing relation of the various goroutines and which ones signal to each other
* Should help prevent throttling loop interfering with the other loading/publishing jobs of the batcher (i.e. should increase its throughput when under load). 
* Should reduce the chance that an active sequencer change results in a delay to the throttling signal
* Be very interesting to know the impact of this change on batcher throughput. We need some acceptance testing / criteria around this.


---
TODO 
- [x] to remove the throttlingLoop config variable
- [x] add an "event" for active sequencer change
- [x] stress test this batcher using kurtosis and tx fuzzing
- [ ] prepare release notes / docs on how THROTTLE_INTERVAL is deprecated

geoknee committed:
op-batcher: overhaul throttling, reading and writing loops

* no longer evaluate throttling conditions on a ticker
* break main loop into "reading" and "writing"
* reading loop signals to throttling and writing when ALL blocks are loaded (this could be done in a more fine grained way, even when each block is loaded)
* these run concurrently

geoknee committed:
improvements

geoknee committed:
readloop unblocks writing loop once, and then writing loop goes forever (pausing when it runs out of data)

tests pass but timeout due to bad shutdown

geoknee committed:
rename to prevent shadowing

geoknee committed:
allow for receipt handling when receipt and err are both nil

geoknee committed:
split shutdownCtx into producers and consumers

geoknee committed:
throttling loop ranges over a channel, does not take a context

geoknee committed:
processReceiptsLoop ranges over a channel, does not take a context

geoknee committed:
unify wait groups

geoknee committed:
unify contexts

geoknee committed:
writeLoop uses a ticker instead of a sleep

geoknee committed:
WIP: add throughput test for batcher

geoknee committed:
make txQueue local, not global state

geoknee committed:
reduce diff

geoknee committed:
reduce diff further

geoknee committed:
make pendingBytesUpdated a local, not global

geoknee committed:
read and write loop communicate with a channel

no more ticker required

geoknee committed:
rename

geoknee committed:
rename

geoknee committed:
abstract promptLoop

geoknee commented on op-e2e/system/da/throughput_test.go:
```diff
@@ -0,0 +1,73 @@
+	"github.com/stretchr/testify/require"
+)
+
+// TestBatcherThroughput tests the throughput of the batcher by creating a large amount of L2 data
+// before starting the batcher. This causes a backlog which must be cleared quickly by the batcher.
+func TestBatcherThroughput(t *testing.T) {
```
Not sure about adding this test, yet. So far been unable to show any difference in throughput with the changes from this PR.

geoknee committed:
rename

geoknee committed:
update readme

geoknee committed:
add diagram to readme

geoknee committed:
remove test

I'm not sure it is adding any value. May return to it in future

geoknee committed:
sendToThrottlingLoop uses mutex

geoknee committed:
harmonize "*Loop returning" logs

geoknee committed:
tidy

geoknee committed:
remove TODO

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -465,68 +494,54 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
+func (l *BatchSubmitter) throttlingLoop(wg *sync.WaitGroup, pendingBytesUpdated chan int64) {
+	defer wg.Done()
 	l.Log.Info("Starting DA throttling loop")
-	ticker := time.NewTicker(l.Config.ThrottleInterval)
-	defer ticker.Stop()
```
By removing the ticker, we only "retry" RPC calls to the sequencer when we get signalled by the `readLoop`, rather than on the next tick. And when we get that signal, we reevaluate the need to make the RPC calls and may not make them after all (so it's not really a retry). We could install some generic retry logic in here, to retry immediately (a limited number of times) if there is an RPC error.

sebastianst replied:
Yes that makes sense. One idea, we create a `time.Timer` in the throttling loop on which we also `select` and then inside `updateParams` at the end where there are the two retry cases, we `Reset()` the timer with some sane duration, like 10s. This way, the timer only fires (again) when reset by a retry condition.

geoknee replied:
I went ahead with that here cc26ac9f418adc2f9e1bb98d534814b250752d96

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -635,32 +645,31 @@ func (l *BatchSubmitter) waitNodeSync() error {
 			}
-			return
-		}
-		if time.Since(start) > maxDuration {
-			l.Log.Warn("Aborting state publishing, max duration exceeded")
-			return
```
We don't need to return from this loop when we "run out of time" any more, since throttling and reading blocks now happen concurrently.

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -158,19 +154,27 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
+
+	l.wg.Add(3)
+	go l.receiptsLoop(l.wg, receiptsCh)                                   // ranges over receiptsCh channel
+	go l.writeLoop(l.shutdownCtx, l.wg, receiptsCh, blocksLoaded)         // ranges over blocksLoaded, sends on receiptsCh + closes it when done
```
(from call with @sebastianst)

The write loop does not actually directly send on the receiptsCh. It is responsible for spawning _multiple senders_ on that channel. Therefore there is a risk that we try to send on a closed channel (panic). We should check that we don't close that channel before all producers are done (I think we do via `txQueue.Wait()`.

geoknee replied:
We should reflect this in the diagram in the readme. And code comments should be made more accurate.

sebastianst commented on packages/contracts-bedrock/lib/automate:
probably unwanted diff?

geoknee replied:
[1c107e1](https://github.com/ethereum-optimism/optimism/pull/14244/commits/1c107e127b99e86b951e1465b3410089eba59a84)

sebastianst commented on op-batcher/readme.md:
```diff
@@ -29,23 +29,52 @@ The philosophy behind the current architecture is:
+architecture-beta
+    group bs(server)[BatchSubmitter]
+
+    service readLoop(server)[readLoop] in bs
+    service writeLoop(server)[writeLoop] in bs
```
I have a suggestion to rename the loops to make it clearer what they do, and avoid the confusion why the readLoop is producing data for two other loops:
* `readLoop` -> `loaderLoop` or `blockLoaderLoop` or so
* `writeLoop` -> `publishLoop` or so

Also in code of course, not only here

geoknee replied:
Yeah those names work and are in better harmony with existing nomenclature in this component.

geoknee replied:
[7196f19](https://github.com/ethereum-optimism/optimism/pull/14244/commits/7196f1919c30780251887b50253a6f4644dfde04)

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -29,6 +29,8 @@ type ActiveL2RollupProvider struct {
 	rollupIndex         int
 	clientLock          *sync.Mutex
+
+	ActiveProviderChanged *sync.Cond
```
It's discouraged to use `sync.Cond`, it has a complex usage pattern. Why not just a `changedHook func()` that can be set at construction? That's more versatile, in particular could also contain actions on a `sync.Cond` or a channel or any other synchronization method. In our case, we could just send a signal on a `chan struct{}` that the throttling loop listens on.

geoknee replied:
Yeah I agree. This was actually just something I explored and didn't mean to commit.

geoknee replied:
[2988863](https://github.com/ethereum-optimism/optimism/pull/14244/commits/2988863da57b37f4d3cfbef36acbf11037767a0e)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -576,19 +597,10 @@ func (l *BatchSubmitter) throttlingLoop(ctx context.Context) {
 		}
 	}
 
-	cachedPendingBytes := int64(0)
-	for {
-		select {
-		case <-ticker.C:
-			updateParams(int64(cachedPendingBytes))
-		case pendingBytes := <-l.pendingBytesUpdated:
-			cachedPendingBytes = pendingBytes
-			updateParams(pendingBytes)
-		case <-ctx.Done():
-			l.Log.Info("DA throttling loop done")
-			return
-		}
+	for pendingBytes := range pendingBytesUpdated {
+		updateParams(pendingBytes)
 	}
```
if we add a signal hook to the active endpoint provider (AEP), we could receive the signal here, and then reuse the latest pendingBytes value that we locally store in a variable, like
```go
for {
		select {
		case <-activeChanged:
			updateParams(int64(cachedPendingBytes))
		case pendingBytes, ok := <-l.pendingBytesUpdated:
			if !ok {return;} // or break
			cachedPendingBytes = pendingBytes
			updateParams(pendingBytes)
}
```
where `activeChanged` would be the channel on wich the AEP signals. So we'd basically replace the ticker by the AEP signal and shut down the routine by a signal that the pending bytes update channel got closed.

geoknee replied:
Yeah I like this. Will need to put back the `cachedPendingBytes` var, but it makes sense to keep it so we can prod the active sequencer with up-to-date information.

geoknee replied:
[2988863](https://github.com/ethereum-optimism/optimism/pull/14244/commits/2988863da57b37f4d3cfbef36acbf11037767a0e)

geoknee committed:
rip out ThrottleInterval config var

geoknee committed:
add activeSequencerChanged channel and wire it up to onActiveSequencerChanged hook in active rollup provider

geoknee committed:
set callback at runtime, not in constructor

geoknee committed:
reinstate startup order

geoknee committed:
tidy

geoknee committed:
remove dead code

geoknee committed:
remove unintentional change

geoknee committed:
rename readLoop to blockLoadingLoop and writeLoop to publishingLoop

geoknee committed:
improve diagram

geoknee committed:
replace ThrottleInterval with ThrottleThreshold as enabling var

geoknee committed:
remove onActiveProviderChanged arg from constructors

geoknee committed:
protect callback invocation with nullity check

geoknee committed:
lint

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -158,22 +158,31 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
+	pendingBytesUpdated := make(chan int64)
+	blocksLoaded := make(chan struct{})
 
 	// DA throttling loop should always be started except for testing (indicated by ThrottleInterval == 0)
```
```suggestion
	// DA throttling loop should always be started except for testing (indicated by ThrottleThreshold == 0)
```

sebastianst commented on op-batcher/batcher/config.go:
```diff
@@ -96,11 +96,8 @@ type CLIConfig struct {
 
-	// ThrottleInterval is the interval between notifying the block builder of the latest DA throttling state, or 0 to
+	// ThrottleThreshold is the number of pending bytes beyond which the batcher will start throttling future bytes. Set to 0 to
 	// disable notifications entirely (only recommended for testing).
```
```suggestion
	// ThrottleThreshold is the number of pending bytes beyond which the batcher will start throttling future bytes. Set to 0 to
	// disable sequencer throttling entirely (only recommended for testing).
```
Clearer that throttling isn't an internal process, but works on the sequencer.

You may also need to update the `flags.ThrottleThresholdFlag` description.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -158,22 +158,31 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
-	go l.mainLoop(l.shutdownCtx, receiptsCh, cancelReceiptsLoopCtx, cancelThrottlingLoopCtx) // sends on receiptsCh
+
+	l.wg.Add(3)
+	go l.receiptsLoop(l.wg, receiptsCh)                                           // ranges over receiptsCh channel
+	go l.publishingLoop(l.shutdownCtx, l.wg, receiptsCh, blocksLoaded)            // ranges over blocksLoaded, spawns routines which send on receiptsCh. Closes receiptsCh when done.
```
I just realize a change in behavior. Before, the publishing function would always execute even if no blocks were loaded. Now, that only happens if blocks were loaded. While this is almost always the case, there are edge cases where the old behavior was better. E.g. if there's a temporary connection loss to the sequencer and no blocks could be loaded, there may still be channels timing out in the meantime, which should now be published. So we should find a way to make sure the publish function is still called periodically even if no blocks were loaded, so it's not only driven by the loader loop. I think it does make sense to generally run it directly after blocks were loaded, but still make sure, maybe using an internal backup ticker, that it's run periodically if the blocks loaded signal isn't coming in for a while. Maybe we could reset that internal backup ticker after every publish call to 2*update-interval, or so.

geoknee replied:
Another option is to just always signal the publishing loop after trying to load blocks. Then it will always happen on the `PollInterval`, like before, and we don't need to introduce more tickers and intervals.

geoknee replied:
Went for that here 69858c97b5633b3ffba62d159664a47f727ead1c

geoknee replied:
We might want to consider renaming the `blocksLoadedCh` to something more generic like `publishingLoopTriggerCh`, to reflect the fact that we may signal even when no blocks are loaded.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -29,6 +29,14 @@ type ActiveL2RollupProvider struct {
 	rollupIndex         int
 	clientLock          *sync.Mutex
+
+	// callback function to be called when the active provider changes
+	onActiveProviderChanged func()
+}
+
+// SetOnActiveProviderChanged sets the callback function to be called when the active provider changes
+func (a *ActiveL2RollupProvider) SetOnActiveProviderChanged(onActiveProviderChanged func()) {
+	a.onActiveProviderChanged = onActiveProviderChanged
```
Would be nice to extend one test of the active rollup provider to test that the hook works as intended.

geoknee replied:
70b5e9cc0efd651daf5e59dcdd17c7a163631bed

sebastianst commented on op-batcher/readme.md:
```diff
@@ -24,38 +27,90 @@ The architecture of this batcher implementation is shown on the left side of the
+
+The `throttlingLoop` which
+
+1. Waits for a signal from the `blockLoadingLoop`
```
And also from the active endpoint provider, if more than one seq endpoints were provided.

geoknee replied:
[ec48011](https://github.com/ethereum-optimism/optimism/pull/14244/commits/ec480119d8b14103be503ce1c81b64e1e7fcb73e)

sebastianst commented on op-batcher/batcher/service.go:
```diff
@@ -141,6 +139,19 @@ func (bs *BatcherService) initFromCLIConfig(ctx context.Context, version string,
+			}
+		}
+		provider.ActiveL2RollupProvider.SetOnActiveProviderChanged(cb)
+	}
```
I think it would be nicer to move this into `initRPCClients` inside the branch that creates the `NewActiveL2EndpointProvider`. And only set the callback if throttling is active (value >0).

geoknee replied:
This is tricky because we need the driver to be initialised first -- and things are started up in quite a delicate order. Added the condition about throttling being enabled here ab3266562952656fa7fd0699ddd386836792030c

sebastianst replied:
I see... I have an idea: https://github.com/ethereum-optimism/optimism/pull/14456 Feel free to merge or discard. This manages to move the initialization code into one place, at the expense of introducing a pattern where the init functions can return optional `DriverSetupOption`s.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -125,9 +122,11 @@ func NewBatchSubmitter(setup DriverSetup) *BatchSubmitter {
-		channelMgr:  state,
+		DriverSetup:            setup,
+		channelMgr:             state,
+		activeSequencerChanged: make(chan struct{}),
```
Wdyt about giving this a buffer of size 1 so that a try-send in the hook doesn't miss sending a signal if the throttling loop is busy in that moment? There may be edge cases where we'd miss updating a fresh active sequencer because of the try-send.

geoknee replied:
Yes I like this 31c9c76a46ef5b874037a21fd8c0821269590714

sebastianst commented on op-batcher/batcher/service.go:
```diff
@@ -141,6 +139,19 @@ func (bs *BatcherService) initFromCLIConfig(ctx context.Context, version string,
+		cb := func() {
+			select {
+			case bs.driver.activeSequencerChanged <- struct{}{}:
+			default:
```
see other comment, because we make a try-send here, I think we should give the channel a buffer size of 1.

geoknee replied:
Resolved here, added the buffer 31c9c76a46ef5b874037a21fd8c0821269590714

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -382,6 +385,34 @@ const (
+
+	// notify the throttling loop it may be time to initiate throttling without blocking
+	select {
+	case pendingBytesUpdated <- pendingBytes:
+	default:
```
we should make sure that this channel also has a buffer of size 1 so this signal doesn't get lost.

geoknee replied:
7769b7234ccf090cf2e877f4da93bd530aeba54c

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -635,32 +664,31 @@ func (l *BatchSubmitter) waitNodeSync() error {
+		case <-ctx.Done():
 			return
-		}
+		default:
```
You could move the code inside the `default` case just after the `select` to save one indentation.

Why does the function now receive a `ctx`, but passes the `l.killCtx` to the actual `l.publishTxToL1` call? In isolation, this looks confusing.

geoknee replied:
0f27d9e4f36dec7f2ac3bf17de37160944f14f31 and 25785651080390fae96ee3d6d2581dca5ad11883

sebastianst replied:
I'm wondering now, why do we even pass the kill but not the shutdown context to the queue?

geoknee replied:
From what I can tell, there are two shutdown modes for our op-services (including the batcher). Graceful shutdown (the first interrupt signal) or forced shutdown / kill (the second interrupt signal). 

On graceful shutdown, we wait for in flight transactions from the queue to confirm or reject. On forced shutdown, we do not. 

Our e2e tests appear to use a forced shutdown.

Happy to dig into this further, but I think it is somewhat outside scope or could at least be tackled in a followup.

sebastianst replied:
ok that makes sense, to wait for in-flight txs. Let's leave it where it is then.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -465,68 +505,53 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
+				if err := l.loadBlocksIntoState(ctx, blocksToLoad.start, blocksToLoad.end); errors.Is(err, ErrReorg) {
 					l.Log.Warn("error loading blocks, clearing state and waiting for node sync", "err", err)
 					l.waitNodeSyncAndClearState()
 					continue
+				} else {
+					l.sendToThrottlingLoop(pendingBytesUpdated) // we have increased the pending data. Signal the throttling loop to check if it should throttle.
+					l.signalPublishingLoop(blocksLoadedCh)      // signal the write loop that blocks have been loaded
 				}
```
because of the `continue`, you could move this out of the `else` branch, like the publish call before.

geoknee replied:
69858c97b5633b3ffba62d159664a47f727ead1c

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -158,22 +158,31 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
-	receiptsLoopCtx, cancelReceiptsLoopCtx := context.WithCancel(context.Background())
-	throttlingLoopCtx, cancelThrottlingLoopCtx := context.WithCancel(context.Background())
+
+	l.txpoolMutex.Lock()
+	l.txpoolState = TxpoolGood
+	l.txpoolMutex.Unlock()
```
Do we even need to lock the mutex here? I think it's setup code and the other routines read/writing it are started later.

geoknee replied:
488352e57bee620a535654cbb3b38b50019b2286

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -382,6 +385,34 @@ const (
+	}
+}
+
+// signalPublishingLoop sends the current pending bytes to the throttling loop.
+// It is not blocking, no signal will be sent if the channel is full.
+func (l *BatchSubmitter) signalPublishingLoop(blocksLoaded chan struct{}) {
+	// notify the publishingLoop in a non blocking way
+	select {
+	case blocksLoaded <- struct{}{}:
+	default:
+	}
+}
```
Note that this doesn't need to be a method on the `BatchSubmitter`, you don't use any of its state ;) could just be a generic `trySignal` function.

geoknee replied:
2c22d9fe05c7a4eb0a6107c0a9579d447fa0adaf

geoknee committed:
Apply suggestions from code review

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

geoknee committed:
remove ThrottelInterval from flags

geoknee committed:
docs: mention active sequencer signal

geoknee committed:
only attach callback if throttling is enabled

geoknee committed:
increase buffer of activeSequencerChanged

means that we buffer a single event if the throttlingLoop is busy, even when using a try-send

geoknee committed:
add buffer to pendingBytesUpdated

see previous commit

geoknee committed:
reduce indentation

geoknee committed:
remove dangling ref

geoknee committed:
pass killCtx to publishingLoop and avoid accessing this context globally

geoknee committed:
remove continue and always signal publishing loop

geoknee committed:
remove unecessary mutex wrangle

geoknee committed:
replace signalPublishingLoop method with trySignal fn

geoknee committed:
extend TestRollupProvider_FailoverOnInactiveSequencer to cover callback fn

geoknee committed:
use retryTimer in throttlingLoop

if RPC calls fail, they will be retried after retryInterval (or before, if another event triggers updateParams)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -465,68 +502,57 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
-
-			l.publishStateToL1(queue, receiptsCh, daGroup, l.Config.PollInterval)
-
+			trySignal(blocksLoadedCh) // always signal the write loop to ensure we periodically publish even if we aren't loading blocks
```
maybe we rename the channel to something like `publishSignal`?

geoknee replied:
[06a9228](https://github.com/ethereum-optimism/optimism/pull/14244/commits/06a9228837b8658ce974ac69509f3ea1982582ab)

sebastianst committed:
op-batcher: improve active-seq-changed signalling setup

geoknee committed:
Merge pull request #14456 from ethereum-optimism/seb/gk-spike-throttle

op-batcher: improve active-seq-changed signalling setup

geoknee committed:
remove unused state var

geoknee committed:
rename blocksLoaded channel to publishSignal channel
